From 3e672bd7d34400bad24deebf601ee3d67fad3e19 Mon Sep 17 00:00:00 2001
From: Alexey Makhalov <alexey.makhalov@broadcom.com>
Date: Thu, 11 Sep 2025 17:26:16 +0000
Subject: [PATCH 14/16] Add ghash accelerator

Prepare glue C files to be included into the canister binary. Move to
wrapper calls. Remove modules related stuff and use fcw_ functions.

.S file: Disable retpoline and return sites and use "ret;int3" for
straight-line-speculation mitigation.

crypto/Makefile: include corresponding .o files.

Signed-off-by: Alexey Makhalov <alexey.makhalov@broadcom.com>
---
 arch/x86/crypto/ghash-clmulni-intel_asm.S  |  5 +++
 arch/x86/crypto/ghash-clmulni-intel_glue.c | 52 +++++++---------------
 crypto/Makefile                            |  4 +-
 3 files changed, 25 insertions(+), 36 deletions(-)

diff --git a/arch/x86/crypto/ghash-clmulni-intel_asm.S b/arch/x86/crypto/ghash-clmulni-intel_asm.S
index 99cb983de..3a50fee3a 100644
--- a/arch/x86/crypto/ghash-clmulni-intel_asm.S
+++ b/arch/x86/crypto/ghash-clmulni-intel_asm.S
@@ -29,6 +29,11 @@
 #define BSWAP	%xmm5
 #define IN1	%xmm6
 
+#if defined(RET)
+#undef RET
+#define RET ret;int3
+#endif
+
 .text
 
 /*
diff --git a/arch/x86/crypto/ghash-clmulni-intel_glue.c b/arch/x86/crypto/ghash-clmulni-intel_glue.c
index 41bc02e48..4360991c1 100644
--- a/arch/x86/crypto/ghash-clmulni-intel_glue.c
+++ b/arch/x86/crypto/ghash-clmulni-intel_glue.c
@@ -7,23 +7,20 @@
  *   Author: Huang Ying <ying.huang@intel.com>
  */
 
-#include <linux/err.h>
-#include <linux/module.h>
-#include <linux/init.h>
-#include <linux/kernel.h>
-#include <linux/crypto.h>
-#include <crypto/algapi.h>
 #include <crypto/cryptd.h>
 #include <crypto/gf128mul.h>
 #include <crypto/internal/hash.h>
-#include <crypto/internal/simd.h>
-#include <asm/cpu_device_id.h>
-#include <asm/simd.h>
-#include <linux/unaligned.h>
 
 #define GHASH_BLOCK_SIZE	16
 #define GHASH_DIGEST_SIZE	16
 
+extern void fcw_kernel_fpu_begin(void);
+extern void fcw_kernel_fpu_end(void);
+extern bool fcw_x86_match_cpu_pcmul_cpu_id(void);
+extern bool fcw_ghash_do_async(struct cryptd_ahash *tfm);
+int __init ghash_pclmulqdqni_mod_init(void);
+void __exit ghash_pclmulqdqni_mod_exit(void);
+
 void clmul_ghash_mul(char *dst, const le128 *shash);
 
 void clmul_ghash_update(char *dst, const char *src, unsigned int srclen,
@@ -99,7 +96,7 @@ static int ghash_update(struct shash_desc *desc,
 	struct ghash_ctx *ctx = crypto_shash_ctx(desc->tfm);
 	u8 *dst = dctx->buffer;
 
-	kernel_fpu_begin();
+	fcw_kernel_fpu_begin();
 	if (dctx->bytes) {
 		int n = min(srclen, dctx->bytes);
 		u8 *pos = dst + (GHASH_BLOCK_SIZE - dctx->bytes);
@@ -115,7 +112,7 @@ static int ghash_update(struct shash_desc *desc,
 	}
 
 	clmul_ghash_update(dst, src, srclen, &ctx->shash);
-	kernel_fpu_end();
+	fcw_kernel_fpu_end();
 
 	if (srclen & 0xf) {
 		src += srclen - (srclen & 0xf);
@@ -138,9 +135,9 @@ static void ghash_flush(struct ghash_ctx *ctx, struct ghash_desc_ctx *dctx)
 		while (dctx->bytes--)
 			*tmp++ ^= 0;
 
-		kernel_fpu_begin();
+		fcw_kernel_fpu_begin();
 		clmul_ghash_mul(dst, &ctx->shash);
-		kernel_fpu_end();
+		fcw_kernel_fpu_end();
 	}
 
 	dctx->bytes = 0;
@@ -196,8 +193,7 @@ static int ghash_async_update(struct ahash_request *req)
 	struct ghash_async_ctx *ctx = crypto_ahash_ctx(tfm);
 	struct cryptd_ahash *cryptd_tfm = ctx->cryptd_tfm;
 
-	if (!crypto_simd_usable() ||
-	    (in_atomic() && cryptd_ahash_queued(cryptd_tfm))) {
+	if (fcw_ghash_do_async(cryptd_tfm)) {
 		memcpy(cryptd_req, req, sizeof(*req));
 		ahash_request_set_tfm(cryptd_req, &cryptd_tfm->base);
 		return crypto_ahash_update(cryptd_req);
@@ -214,8 +210,7 @@ static int ghash_async_final(struct ahash_request *req)
 	struct ghash_async_ctx *ctx = crypto_ahash_ctx(tfm);
 	struct cryptd_ahash *cryptd_tfm = ctx->cryptd_tfm;
 
-	if (!crypto_simd_usable() ||
-	    (in_atomic() && cryptd_ahash_queued(cryptd_tfm))) {
+	if (fcw_ghash_do_async(cryptd_tfm)) {
 		memcpy(cryptd_req, req, sizeof(*req));
 		ahash_request_set_tfm(cryptd_req, &cryptd_tfm->base);
 		return crypto_ahash_final(cryptd_req);
@@ -255,8 +250,7 @@ static int ghash_async_digest(struct ahash_request *req)
 	struct ahash_request *cryptd_req = ahash_request_ctx(req);
 	struct cryptd_ahash *cryptd_tfm = ctx->cryptd_tfm;
 
-	if (!crypto_simd_usable() ||
-	    (in_atomic() && cryptd_ahash_queued(cryptd_tfm))) {
+	if (fcw_ghash_do_async(cryptd_tfm)) {
 		memcpy(cryptd_req, req, sizeof(*req));
 		ahash_request_set_tfm(cryptd_req, &cryptd_tfm->base);
 		return crypto_ahash_digest(cryptd_req);
@@ -331,17 +325,11 @@ static struct ahash_alg ghash_async_alg = {
 	},
 };
 
-static const struct x86_cpu_id pcmul_cpu_id[] = {
-	X86_MATCH_FEATURE(X86_FEATURE_PCLMULQDQ, NULL), /* Pickle-Mickle-Duck */
-	{}
-};
-MODULE_DEVICE_TABLE(x86cpu, pcmul_cpu_id);
-
-static int __init ghash_pclmulqdqni_mod_init(void)
+int __init ghash_pclmulqdqni_mod_init(void)
 {
 	int err;
 
-	if (!x86_match_cpu(pcmul_cpu_id))
+	if (!fcw_x86_match_cpu_pcmul_cpu_id())
 		return -ENODEV;
 
 	err = crypto_register_shash(&ghash_alg);
@@ -359,15 +347,9 @@ static int __init ghash_pclmulqdqni_mod_init(void)
 	return err;
 }
 
-static void __exit ghash_pclmulqdqni_mod_exit(void)
+void __exit ghash_pclmulqdqni_mod_exit(void)
 {
 	crypto_unregister_ahash(&ghash_async_alg);
 	crypto_unregister_shash(&ghash_alg);
 }
 
-module_init(ghash_pclmulqdqni_mod_init);
-module_exit(ghash_pclmulqdqni_mod_exit);
-
-MODULE_LICENSE("GPL");
-MODULE_DESCRIPTION("GHASH hash function, accelerated by PCLMULQDQ-NI");
-MODULE_ALIAS_CRYPTO("ghash");
diff --git a/crypto/Makefile b/crypto/Makefile
index 935c41497..19e8d78a4 100644
--- a/crypto/Makefile
+++ b/crypto/Makefile
@@ -222,7 +222,9 @@ sha256-ssse3-$(CONFIG_AS_SHA256_NI) += sha256_ni_asm.o
 
 sha512-ssse3-y := sha512-ssse3-asm.o sha512-avx-asm.o sha512-avx2-asm.o sha512_ssse3_glue.o
 
-x86-hwaccelerated-y := $(aesni-intel-y) $(sha256-ssse3-y) $(sha512-ssse3-y)
+ghash-clmulni-intel-y := ghash-clmulni-intel_asm.o ghash-clmulni-intel_glue.o
+
+x86-hwaccelerated-y := $(aesni-intel-y) $(sha256-ssse3-y) $(sha512-ssse3-y) $(ghash-clmulni-intel-y)
 
 
 crypto/x86-%.o: arch/x86/crypto/%.c $(recordmcount_source) $(objtool_dep) FORCE
-- 
2.40.4

